File: daily_python/2220_Minimum_Bit_Flips_to_Convert_Number.py
import os
import sys

class Solution:
    def count_one(self, i):
        count = 0
        while i:
            if i % 2 == 1:
                count += 1
            i //= 2
        return count
    
    def minBitFlips(self, start: int, goal: int) -> int:
        return self.count_one(start ^ goal)

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            start = int(lines[0].split('=')[1].strip()[:-1])
            goal = int(lines[1].split('=')[1].strip()[:-1])
            dataset.append((start, goal))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for start, goal in dataset:
        results.append(solution.minBitFlips(start, goal))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')

File: daily/2024/2220_Minimum_Bit_Flips_to_Convert_Number.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int minBitFlips(int start, int goal) {
        return __builtin_popcount(start ^ goal);
    }
};

int main(){
    int start = 10;
    int goal = 7;
    // int start = 3;
    // int goal = 4;
    Solution S;
    cout << S.minBitFlips(start, goal) << endl;
    return 0;
}